src/Build/Evaluation/LazyItemEvaluator.cs

```diff
@@ -516,21 +516,51 @@ private IncludeOperation BuildIncludeOperation(string rootDirectory, ProjectItem
 
         private RemoveOperation BuildRemoveOperation(string rootDirectory, ProjectItemElement itemElement, bool conditionResult)
         {
-            OperationBuilder operationBuilder = new OperationBuilder(itemElement, conditionResult);
+            RemoveOperationBuilder operationBuilder = new RemoveOperationBuilder(itemElement, conditionResult);
 
             ProcessItemSpec(rootDirectory, itemElement.Remove, itemElement.RemoveLocation, operationBuilder);
 
+            // Process MatchOnMetadata
+            if (itemElement.MatchOnMetadata.Length > 0)
+            {
+                string evaluatedmatchOnMetadata = _expander.ExpandIntoStringLeaveEscaped(itemElement.MatchOnMetadata, ExpanderOptions.ExpandProperties, itemElement.MatchOnMetadataLocation);
+
+                if (evaluatedmatchOnMetadata.Length > 0)
+                {
+                    var matchOnMetadataSplits = ExpressionShredder.SplitSemiColonSeparatedList(evaluatedmatchOnMetadata);
+
+                    ImmutableList<string>.Builder metadataPatterns = ImmutableList.CreateBuilder<string>();
+                    foreach (var matchOnMetadataSplit in matchOnMetadataSplits)
+                    {
+                        metadataPatterns.Add(matchOnMetadataSplit);
+                        AddItemReferences(matchOnMetadataSplit, operationBuilder, itemElement.MatchOnMetadataLocation);
+                    }
+                    
+                    foreach (string metadata in metadataPatterns)
+                    {
+                        string metadataExpanded = _expander.ExpandIntoStringLeaveEscaped(metadata, ExpanderOptions.ExpandPropertiesAndItems, itemElement.MatchOnMetadataLocation);
+                        var metadataSplits = ExpressionShredder.SplitSemiColonSeparatedList(metadataExpanded);
+                        operationBuilder.MatchOnMetadata.AddRange(metadataSplits);
+                    }
+                }
+            }
+
+            operationBuilder.MatchOnMetadataOptions = MatchOnMetadataOptions.CaseSensitive;
```

Comment:
nit:
You can save one line and perhaps improve clarity slightly if you make it:

if (!Enum.TryParse(itemElement.MatchOnMetadataOptions, out operationBuilder.MatchOnMetadataOptions)
{
    operationBuilder.MatchOnMetadataOptions = MatchOnMetadataOptions.CaseSensitive;
}
I would read that as "if we try to parse it and fail, this is the default," which is clearer than "here's the default; now let's try to parse it."